core/src/main/java/org/jdbi/v3/core/internal/QualifiedEnumMapperFactory.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.internal;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.jdbi.v3.core.Enums;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.mapper.ColumnMapper;
+import org.jdbi.v3.core.mapper.QualifiedColumnMapperFactory;
+import org.jdbi.v3.core.qualifier.QualifiedType;
+import org.jdbi.v3.core.result.UnableToProduceResultException;
+import org.jdbi.v3.core.statement.StatementContext;
+
+public class QualifiedEnumMapperFactory implements QualifiedColumnMapperFactory {
+    private static final Map<Class<? extends Enum<?>>, ColumnMapper<? extends Enum<?>>> BY_NAME_MAPPER_CACHE = new ConcurrentHashMap<>();
+
+    @Override
+    public Optional<ColumnMapper<?>> build(QualifiedType<?> givenType, ConfigRegistry config) {
+        return Optional.of(givenType.getType())
+            .map(GenericTypes::getErasedType)
+            .filter(Class::isEnum)
+            .flatMap(clazz -> makeEnumArgument((QualifiedType<Enum>) givenType, (Class<Enum>) clazz, config));
+    }
+
+    private static <E extends Enum<E>> Optional<ColumnMapper<?>> makeEnumArgument(QualifiedType<E> givenType, Class<E> enumClass, ConfigRegistry config) {
+        boolean byName = Enums.EnumStrategy.BY_NAME == config.get(Enums.class).findStrategy(givenType, enumClass);
+
+        return Optional.of(byName
+            ? byName(enumClass)
+            : byOrdinal(enumClass));
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <E extends Enum<E>> ColumnMapper<E> byName(Class<E> enumClass) {
+        return (ColumnMapper<E>) BY_NAME_MAPPER_CACHE.computeIfAbsent(enumClass, e -> new EnumByNameColumnMapper<>(enumClass));
+    }
+
+    public static <E extends Enum<E>> ColumnMapper<E> byOrdinal(Class<E> enumClass) {
+        return new EnumByOrdinalColumnMapper<>(enumClass);
+    }
+
+    private static class EnumByNameColumnMapper<E extends Enum<E>> implements ColumnMapper<E> {
+        private final Map<String, E> nameValueCache = new ConcurrentHashMap<>();
+        private final Class<E> enumClass;
+
+        private EnumByNameColumnMapper(Class<E> enumClass) {
+            this.enumClass = enumClass;
+        }
+
+        @Override
+        public E map(ResultSet rs, int columnNumber, StatementContext ctx) throws SQLException {
+            String name = ctx.findColumnMapperFor(String.class)
+                .orElseThrow(() -> new UnableToProduceResultException("a String column mapper is required to map Enums from names", ctx))
+                .map(rs, columnNumber, ctx);
+
+            return name == null ? null : nameValueCache.computeIfAbsent(name.toLowerCase(), lowercased -> getValueByName(enumClass, name, ctx));
```

Comment:
Added a benchmark for enum mapping. It seeds 1000 enums into two tables: one with exact names, and one with randomized case names. The same enum values are inserted into both tables so we're comparing apples to apples.

Variables to test:

- with or without cache
- lowercase cache key or not
- case sensitive (Enum.valueOf) vs insensitive

Without cache, Enum.valueOf() only (baseline):

```
Benchmark                         Mode  Cnt       Score      Error  Units
EnumMapperBenchmark.mapExactCase  avgt    5  137059.724 ± 3079.280  ns/op
```

With cache, Enum.valueOf() only:

```
Benchmark                         Mode  Cnt       Score      Error  Units
EnumMapperBenchmark.mapExactCase  avgt    5  140057.354 ± 2108.761  ns/op
```

Enum.valueOf() maintains a constant directory (map) so it makes sense that adding a second cache actually slowed it down.

Without cache, case insensitive:

```
Benchmark                          Mode  Cnt       Score       Error  Units
EnumMapperBenchmark.mapExactCase   avgt    5  447953.782 ± 10698.905  ns/op
EnumMapperBenchmark.mapRandomCase  avgt    5  816709.142 ± 15131.747  ns/op
```

Adding case-insensitivity (using equalsIgnoreCase) slowed things way down. But..

With cache, case insensitive:

```
Benchmark                          Mode  Cnt       Score      Error  Units
EnumMapperBenchmark.mapExactCase   avgt    5  131139.802 ± 3909.448  ns/op
EnumMapperBenchmark.mapRandomCase  avgt    5  160514.845 ± 1140.261  ns/op
```

So this is actually a little faster for case sensitive, but 17% slower for case insensitive. This might be a reasonable trade-off.

With cache, case insensitive, lowercase cache key:

```
Benchmark                          Mode  Cnt       Score      Error  Units
EnumMapperBenchmark.mapExactCase   avgt    5  179181.546 ± 1043.832  ns/op
EnumMapperBenchmark.mapRandomCase  avgt    5  204252.115 ± 2278.226  ns/op
```

Converting the value to lowercase hurt increased case-sensitive time by 37%, and case-insensitive time by 27%.

If we're committing to support case-insensitive lookups, then using a cache (without altering the key) seems like the way to go.